src/dynamics/mpas/dyn_comp.F90

```diff
@@ -139,13 +175,575 @@ pure function sec_to_hour_min_sec(sec) result(hour_min_sec)
+                        constant_g / (constant_rd * tm_mid_col(k - 1)) * (1.0_kind_r8 + qv_mid_col(k - 1))) * &
+                        exp(-0.5_kind_r8 * (zgrid(k + 1, i) - zgrid(k    , i)) * &
+                        constant_g / (constant_rd * tm_mid_col(k    )) * (1.0_kind_r8 + qv_mid_col(k    )))
+                end do
+
+                rho(:, i) = p_mid_col(:) / (constant_rd * tm_mid_col(:))
+                theta(:, i) = t_mid(:, i) * ((constant_p0 / p_mid_col(:)) ** (constant_rd / constant_cpd))
+            end do
+
+            deallocate(p_mid_col)
+            deallocate(p_sfc)
+            deallocate(qv_mid_col)
+            deallocate(t_mid)
+            deallocate(tm_mid_col)
+
+            nullify(index_qv)
+            nullify(rho)
+            nullify(theta)
+            nullify(scalars)
+
+            ! Because we are injecting data directly into MPAS memory, halo layers need to be updated manually.
+            call mpas_dynamical_core % exchange_halo('rho')
+            call mpas_dynamical_core % exchange_halo('theta')
+        end subroutine set_mpas_state_rho_theta
+
+        !> Set MPAS state `rho_base` (i.e., base state dry air density) and `theta_base` (i.e., base state potential temperature).
+        !> (KCW, 2024-05-21)
+        subroutine set_mpas_state_rho_base_theta_base()
+            integer :: i, k
+            integer :: ierr
+            real(kind_r8), parameter :: t_base = 250.0_kind_r8 ! Base state temperature (K) of dry isothermal atmosphere.
+                                                               ! The value used here is identical to MPAS.
+            real(kind_r8), allocatable :: p_base(:)            ! Base state pressure (Pa) at layer midpoints of each column.
+            real(kind_r8), pointer :: rho_base(:, :) => null()
+            real(kind_r8), pointer :: theta_base(:, :) => null()
+            real(kind_r8), pointer :: zz(:, :) => null()
+
+            call dyn_debug_print('Setting MPAS state "rho_base" and "theta_base"')
+
+            allocate(p_base(pver), stat=ierr)
+            call check_allocate(ierr, 'set_analytic_initial_condition', 'p_base', 'dyn_comp', __LINE__)
+
+            call mpas_dynamical_core % get_variable_pointer(rho_base, 'diag', 'rho_base')
+            call mpas_dynamical_core % get_variable_pointer(theta_base, 'diag', 'theta_base')
+            call mpas_dynamical_core % get_variable_pointer(zz, 'mesh', 'zz')
+
+            ! Set `rho_base` and `theta_base` column by column. This way, peak memory usage can be reduced.
+            do i = 1, ncells_solve
+                do k = 1, pver
+                    ! Derive `p_base` by hypsometric equation.
+                    ! The formulation used here is exact and identical to MPAS.
+                    p_base(k) = constant_p0 * exp(-0.5_kind_r8 * (zgrid(k, i) + zgrid(k + 1, i)) / &
+                        (constant_rd * t_base / constant_g))
```

Comment:
The above rationale holds here, too. The hypsometric equation is found [here](https://github.com/MPAS-Dev/MPAS-Model/blob/v8.2.0/src/core_init_atmosphere/mpas_init_atm_cases.F#L4862-L4863). It is also inlined so that we cannot use it.

However, thanks to your suggestion, I can indeed factor out the equations used here. This commit https://github.com/ESCOMP/CAM-SIMA/pull/267/commits/77cf091c24a89ee229b2ba291f928b85cb9997d5 should hopefully resolve this.